packages/mds-db/events.ts

```diff
@@ -483,3 +494,130 @@ export async function getMostRecentEventByProvider(): Promise<{ provider_id: UUI
   const sql = `select provider_id, max(recorded) from ${schema.TABLE.events} group by provider_id`
   return makeReadOnlyQuery(sql)
 }
+
+// for CMM
+export type TimeRange = {
+  start: Timestamp
+  end: Timestamp
+}
+export interface GetVehicleEventsFilterParams {
+  vehicle_types?: VEHICLE_TYPE[]
+  propulsion_types?: PROPULSION_TYPE[]
+  provider_ids?: UUID[]
+  vehicle_statuses?: VEHICLE_STATUS[]
+  time_range: TimeRange
+  device_or_vehicle_id?: string // Match on device_id or vehicle_id
+  device_ids?: UUID[]
+  event_types?: VEHICLE_EVENT[]
+  geography_ids?: UUID[]
+}
+
+export async function getLatestEventPerVehicle({
+  vehicle_types,
+  propulsion_types,
+  provider_ids,
+  vehicle_statuses,
+  time_range,
+  device_or_vehicle_id,
+  device_ids,
+  event_types,
+  geography_ids
+}: GetVehicleEventsFilterParams): Promise<Recorded<VehicleEvent & Pick<Device, 'vehicle_id'>>[]> {
+  const client = await getReadOnlyClient()
+  const vals = new SqlVals()
+  const exec = SqlExecuter(client)
+
+  const { start: start_time, end: end_time } = time_range
+
+  const conditions: string[] = []
+  const time_range_conditions: string[] = []
+
+  if (!isTimestamp(start_time)) {
+    throw new Error(`invalid start_time ${start_time}`)
+  } else {
+    time_range_conditions.push(`e.timestamp >= ${vals.add(start_time)}`)
+  }
+
+  if (!isTimestamp(end_time)) {
+    throw new Error(`invalid end_time ${end_time}`)
+  } else {
+    time_range_conditions.push(`e.timestamp <= ${vals.add(end_time)}`)
+  }
+
+  if (vehicle_types) {
+    conditions.push(`d.type = ANY (${vals.add(vehicle_types)})`)
+  }
+
+  if (propulsion_types) {
+    conditions.push(`d.propulsion && (${vals.add(propulsion_types)})`)
+  }
+
+  // you can't use non-UUID values to compare to ::UUID types in the database.
+  if (device_or_vehicle_id) {
+    if (isUUID(device_or_vehicle_id)) {
+      conditions.push(`(d.device_id = ${vals.add(device_or_vehicle_id)})`)
+    } else {
+      conditions.push(`(d.vehicle_id = ${vals.add(device_or_vehicle_id)})`)
+    }
+  }
+
+  if (vehicle_statuses) {
+    const vehicle_event_types = vehicle_statuses.map(s => Object.values(STATUS_EVENT_MAP[s])).flat()
+    conditions.push(`e.event_type = ANY (${vals.add(vehicle_event_types)})`)
+  }
+
+  if (event_types) {
+    conditions.push(`e.event_type = ANY (${vals.add(event_types)})`)
+  }
+
+  if (provider_ids) {
+    if (!provider_ids.every(isUUID)) {
+      throw new Error(`invalid provider_ids: ${provider_ids}`)
+    } else {
+      conditions.push(`e.provider_id = ANY (${vals.add(provider_ids)})`)
+    }
+  }
+
+  if (device_ids) {
+    if (!device_ids.every(isUUID)) {
+      throw new Error(`invalid device_ids ${device_ids}`)
+    } else {
+      conditions.push(`e.device_id = ANY (${vals.add(device_ids)})`)
```

Comment:
I take it TypeORM isn't a thing in these parts? Getting some PTSD here with SQL string concatenation.

Reply:
Yeahhh, in the old 0.4 code we only had TypeORM for migrations, never used it for access methods.